### PR TITLE
#2246: Fix copying of sample sheet

### PR DESCRIPTION
**Pull request by seallard**
## Description
This PR fixes the incorrect copy destination for the samplesheet in the new flow cell post processing flow. Part of project https://github.com/Clinical-Genomics/project-planning/issues/482.

### Fixed
- Sample sheet copy destination in demultiplexed flow cell directory

### How to test

- [ ] Do `cg demultiplex finish temporary 20230530_LH00202_0005_B2257WYLT3`

### Expected test outcome

- [ ] Check that the sample sheet is copied into the demux flow cell directory

### This [version](https://semver.org/) is a
- [ ] **PATCH** - when you make backwards compatible bug fixes or documentation/instructions

### CODE REVIEW

**Review by ChrOertlin — APPROVED** [review]
Nice catch!

**Comment by seallard** [comment]
Tested in stage with: 
```
cg demultiplex finish temporary 20230530_LH00202_0005_B2257WYLT3
```
Which completes successfully and copies the sample sheet to the flow cell directory from the run directory.

**Review by diitaz93 — COMMENTED** [review]
I will finish the review in a bit :)

File: tests/conftest.py [hunk 1]
```diff
@@ -2705,11 +2705,29 @@ def flow_cell_directory_name_demultiplexed_with_bcl_convert(
...
+    return dir_path
+
+
+@pytest.fixture(name="flow_cell_runs_directory_with_bcl2fastq")
+def fixture_flow_cell_runs_directory_with_bcl2fastq(
+    flow_cell_runs_dir: Path,
+) -> Path:
+    """Return a directory with flow cell runs with bcl2fastq."""
+    dir_path = Path(flow_cell_runs_dir, "hiseq")
+    return dir_path
```
Comment (diitaz93) [inline tests/conftest.py:2733]: These two fixtures already exist. Their names are `hiseq_dir` and `novaseq_6000_dir`defined in lines 1120 and 1126 respectively

File: tests/conftest.py [hunk 2]
```diff
@@ -2705,11 +2705,29 @@ def flow_cell_directory_name_demultiplexed_with_bcl_convert(
     return f"230504_A00689_0804_B{flow_cell_name_demultiplexed_with_bcl_convert}"
 
 
-@pytest.fixture
-def demultiplexed_flow_cells_directory(tmp_path) -> Path:
+@pytest.fixture(name="demultiplexed_flow_cells_directory")
```
Comment (diitaz93) [inline tests/conftest.py:2708]: I would like to distinguish between directories created using pytest's tmp_path/tmp_dir and directories in `tests/fixtures`. Could you include this in the name? Maybe `demultiplexed_flow_cells_tmp_directory`.

**Review by diitaz93 — APPROVED** [review]
Looks good 👍

File: cg/meta/demultiplex/demux_post_processing.py [hunk 3]
```diff
@@ -68,22 +68,26 @@ def set_dry_run(self, dry_run: bool) -> None:
...
         )
 
+        if not sample_sheet_destination.exists():
+            LOG.debug(
+                f"Copy sample sheet {sample_sheet_source} from flow cell to demuxed result dir {sample_sheet_destination}"
+            )
+            shutil.copy(
+                sample_sheet_source.as_posix(),
+                sample_sheet_destination.as_posix(),
+            )
```
[the remark this answers is not in the document]
Comment (diitaz93) [reply cg/meta/demultiplex/demux_post_processing.py:89]: I guess there is a possibility. I just wonder if there is already a sample sheet the user should know, but maybe is not that necessary

File: tests/conftest.py (hunk 1 above)
Comment (ChrOertlin) [reply tests/conftest.py:2733]: Thanks! I get lost in the fixtures, fixed it.

File: tests/conftest.py (hunk 2 above)
Comment (ChrOertlin) [reply tests/conftest.py:2708]: Yes, I agree. I changed the name

File: cg/meta/demultiplex/demux_post_processing.py [hunk 4]
```diff
@@ -68,22 +68,26 @@ def set_dry_run(self, dry_run: bool) -> None:
...
-        shutil.copy(
-            sample_sheet_path.as_posix(),
-            target_sample_sheet_path.as_posix(),
+        sample_sheet_destination: Path = Path(
+            sample_sheet_destination_directory, DemultiplexingDirsAndFiles.SAMPLE_SHEET_FILE_NAME
         )
 
+        if not sample_sheet_destination.exists():
+            LOG.debug(
+                f"Copy sample sheet {sample_sheet_source} from flow cell to demuxed result dir {sample_sheet_destination}"
```
Comment (seallard) [inline cg/meta/demultiplex/demux_post_processing.py:84]: Might want to change the log message to the destination directory instead
```suggestion
                f"Copy sample sheet {sample_sheet_source} from flow cell to demuxed result dir {sample_sheet_destination_directory}"
```

File: tests/conftest.py [hunk 5]
```diff
@@ -2705,8 +2705,8 @@ def flow_cell_directory_name_demultiplexed_with_bcl_convert(
     return f"230504_A00689_0804_B{flow_cell_name_demultiplexed_with_bcl_convert}"
 
 
-@pytest.fixture
-def demultiplexed_flow_cells_directory(tmp_path) -> Path:
+@pytest.fixture(name="demultiplexed_flow_cells_tmp_directory")
+def fixture_demultiplexed_flow_cells_directory(tmp_path) -> Path:
```
Comment (ChrOertlin) [inline tests/conftest.py:2709]:
```suggestion
def fixture_demultiplexed_flow_cells_tmp_directory(tmp_path) -> Path:
```

File: cg/meta/demultiplex/demux_post_processing.py [hunk 6]
```diff
@@ -118,23 +123,29 @@ def finish_flow_cell_temp(self, flow_cell_directory_name: str) -> None:
...
+
+        bcl_converter: str = get_bcl_converter_name(flow_cell_directory)
 
         parsed_flow_cell: FlowCellDirectoryData = parse_flow_cell_directory_data(
-            flow_cell_directory=flow_cell_directory_path,
+            flow_cell_directory=flow_cell_directory,
             bcl_converter=bcl_converter,
         )
 
+        self.copy_sample_sheet(
```
Comment (seallard) [inline cg/meta/demultiplex/demux_post_processing.py:137]: Maybe `copy_sample_sheet` belongs in the utils module? Something to consider

**Review by seallard — COMMENTED** [review]
I can't approve my own PR, but looks good 👍 
Just one thought: should we raise an exception and abort the flow if we are unable to copy the sample sheet?

File: cg/meta/demultiplex/demux_post_processing.py (hunk 4 above)
Comment (ChrOertlin) [reply cg/meta/demultiplex/demux_post_processing.py:84]: Yes, sounds good!

File: cg/meta/demultiplex/demux_post_processing.py (hunk 6 above)
Comment (ChrOertlin) [reply cg/meta/demultiplex/demux_post_processing.py:137]: I moved the function to utils in this pr now

File: cg/meta/demultiplex/demux_post_processing.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (ChrOertlin) [reply cg/meta/demultiplex/demux_post_processing.py]: Yes, I think so too! I added a warning

**Comment by ChrOertlin** [comment]
> I can't approve my own PR, but looks good 👍 Just one thought: should we raise an exception and abort the flow if we are unable to copy the sample sheet?
> 
> I guess the copy sample sheet function already will fail and raise a file not found exception. Maybe we should try/catch that exception and log a message and then abort

Yes, I added a try and except in the copy sample sheet function what also reports the source and target.

**Comment by ChrOertlin** [comment]
stage:
Logging deploy ...
Getting deployer... done.
Getting last commit message and SHA... done.
Getting version of deploy scripts... /home/christian.oertlin
done.
Log deploy... done.
cg, version 38.3.5



Prod:
Logging deploy ...
Getting deployer... done.
Getting last commit message and SHA... done.
Getting version of deploy scripts... /home/christian.oertlin
done.
Log deploy... done.
cg, version 38.3.5

**Pull request merged**